# Introduction

MicroExonator is a fully-integrated computational pipeline that allows for systematic de novo discovery and quantification of microexons using raw RNA-seq data for any organism with a gene annotation. Compared to other available methods MicroExonator is more sensitive for discovering smaller microexons and it provides higher specificity for all lengths. Moreover, MicroExonator provides integrated downstream comparative analysis between cell types or tissues using [Whippet](https://github.com/timbitz/Whippet.jl) ([Sterne-Weiler et al. 2018](https://doi.org/10.1016/j.molcel.2018.08.018)). As a proof of principle MicroExonator  identified X novel microexons in Y RNA-seq samples from mouse early development to provide a systematic characterization based on time and tissue specificity.


# Install

Start by cloning MicroExonator

    git clone https://github.com/geparada/MicroExonator

Install [Miniconda 2](https://docs.conda.io/en/latest/miniconda.html)

    wget https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh
    chmod +x Miniconda2-latest-Linux-x86_64.sh
    ./Miniconda2-latest-Linux-x86_64.sh -b -p /cvmfs/softdrive.nl/$USER/Miniconda2

Finally, create an enviroment to run [snakemake](https://snakemake.readthedocs.io/en/stable/)

    conda create -n snakemake snakemake python=3.6 pandas cookiecutter

# Configure

Before running MicroExonator you need to have certain files in the `MicroExonator/` directory. First, you need to create a `config.yaml`, which should contain the path of the input files and certain paramethers. The format for the `config.yaml` file is shown below:

    Genome_fasta : /path/to/danRer11.fa
    Gene_anontation_bed12 : /path/to/danRer11.ensembl.bed12
    GT_AG_U2_5 : /path/to/danRer11_GT_AG_U2_5.good.matrix
    GT_AG_U2_3 : /path/to/Zebrafish/Data/danRer11_GT_AG_U2_3.good.matrix
    conservation_bigwig : /path/to/danRer11.chrom.sizes.bw.sep  
    working_directory : /path/to/Zebrafish/
    ME_DB : /path/to/ME_DB.bed
    ME_len : 30

Here:

* `Genome_fasta` is a [multifasta](http://www.metagenomics.wiki/tools/fastq/multi-fasta-format) file containg the cromosomes. 
* `Gene_anontation_bed12` is a BED file containing the transcript annotation. A collection of these files can be found at [UCSC genome browser](http://genome.ucsc.edu/cgi-bin/hgTables). 
* `GT_AG_U2_5` and `GT_AG_U2_5` are splice site PWMs that can be from [SpliceRack](http://katahdin.cshl.edu/SpliceRack/poster_data.html) (as this server is currently down, we provisionally provide the PWM for human and mouse), but if you do not have these PWM for the specie for which you want to conduct the analysis, you can leave it as `NA` and MicroExonator will generate these PWMs internaly.
* `conservation_bigwig` is a bigwig that contain genome-wide conservation scores generated by Pylop or PhastCons, which can be downloaded from UCSC genome browser for some species. If you do not have a `conservation_bigwig` file, you need to create an bigwig which have 0 values for every position of your genome assembly.
* `working_directory` is the path to the MicroExonator folder that you are working with
* `ME_len` is microexon maximum lenght. It should be set up as 30.

To imput the RNA-seq data, you need to either create a `desing.tsv` (for fastq.gz files that are locally in your machine) and/or `NCBI_accession_list.txt`(for SRA accession names). You can find examples of these files inside the Examples folder. 

Finnaly, if you are working in a high performace computing machine, then is very likely that you need to submit jobs to queue systems such as lsf, qsub, SLURM, etc. To make MicroExonator work with these queue systems, you need to create a `cluster.json` file. We currently provide in the Examples folder a `cluster.json` file to run MicroExonator with lsf. To adapt MicroExonator to other quequing systems please see [SnakeMake documentation](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html?highlight=cluster.json#cluster-configuration).


# Run

If you are working remotely or even in your own machine, we highly recommend create an screen before running MicroExonator

    screen -S session_name

Then activate snakemake enviroment

    source activate snakemake

Then run

    snakemake -s MicroExonator.skm  --cluster-config cluster.json --cluster {cluster system params} --use-conda -k  -j {number of parallel jobs}

Notice that you should use `--cluster` only if you are working in a computer cluster that works with queuing systems. We provide an example of `cluster.json` to work with lsf and in that case the cluster system params should be `"bsub -n {cluster.nCPUs} -R {cluster.resources} -c {cluster.tCPU} -G {cluster.Group} -q {cluster.queue} -o {cluster.output} -e {cluster.error} -M {cluster.memory}"`. The number of parallel jobs can have any int number, this depend of your machine work load capacity. 

If you want to process a large dataset, we recommend to run MicroExonator in two stages:

    snakemake -s MicroExonator.skm  --cluster-config cluster.json --cluster {cluster system params} --use-conda -k  -j {number of parallel jobs} discovery
    snakemake -s MicroExonator.skm  --cluster-config cluster.json --cluster {cluster system params} --use-conda -k  -j {number of parallel jobs} quant
    
By doing this, you will optimise disk space, which is often a restrictive resource for running large data sets. 

# Troubleshooting

Before running it is recommended to see if SnakeMake can corretly generate all the steps given your input. For this, you can do a dry-run using `-np` parameters:

    snakemake -s MicroExonator.skm  --cluster-config cluster.json --cluster {cluster system params} --use-conda -k  -j {number of parallel jobs} -np

If the dry-run cannot be iniciated, make sure you are running MicroExonator from inside the folder you cloned from this repository. Make sure you have the right configuration inside `config.yaml`. 

The current version of MicroExonator does not support chromosome names that has `_` or `|`, for example some chromosome names can be `chr1_KZ111v2_alt` and in this case you migth have some errors that will prevent you to complete the run. For now we recommed to replace these caracters by any string you can recognise, for example:

    sed 's/_/SEP/g' genome.fa > genome.fa.sed
    sed 's/_/SEP/g' transcripts.bed > transcripts.bed.sed
    
And then use this modified genome (`genome.fa.sed`) as an input. Future versions of MicroExonator will overcome this issue withouth requering this step.


# Contact

For questions, ideas, feature requests and potential bug reports please contact gp7@sanger.ac.uk.
